_posts/2020-05-30-terraform.md

```diff
@@ -0,0 +1,138 @@
+On one occasion the lead time for a single environment for a new team was as high as 6 weeks.
+This is one of the reasons that at least 1 or 2 spare environments are provisioned in order to provide teams with at least a minimal environment to start working with.
+This of course introduces additional cost.
+
+This is just one of the scenarios which need to be supported.
+Since the choice was made to migrate away from the on-premise solution and onto the public cloud, it was an ideal time to automate these scenarios as well.
+
+## Why use Terraform to supercharge infrastructure provisioning
+
+As mentioned, one of the main drivers to look into automation for infrastructure was the move to the public cloud.
+The reason why Azure was selected, might be discussed in a future blog post.
+
+As Azure was selected, the logical choice for automating the infrastructure provisioning would be ARM templates.
+An attempt was made at setting up the infrastructure with Terraform together with an ARM expert.
+Unfortunately, multiple walls were hit due to the nature of the corporate setup on Azure.
+One of the limitations that were hit, was the inability for ARM to make changes to resources that were provided by a central team.
+More specifically, using a VNET that wasn't managed by ARM to deploy an Azure Kubernetes Service proved to be difficult or even impossible.
+
+Using the CLI and scripts to create the setup was attempted as well, but this didn't fit the vision of a declarative setup for provisioning infrastructure.
+Re-applying the same scripts either broke the setup or created additional, unneeded resources.
+A clear no-go.
+
+As a third option, Terraform was investigated.
+It adhered to the vision of declarative definitions and all the components required by the landscape were available in the Azure Terraform provider.
+Some basic setups were created and it showed great potential for the required setup.
+So Terraform to the rescue!
+
+## How to structure Terraform to allow for a two second tire change
+
+One of the challenges of creating a Terraform setup is to determine a way to structure the code.
+Looking online provided more questions than answers.
+A lot of documentation can be found about how to structure a specific module or how to create modules, but almost no resources discuss how to structure the modules into logical, reusable components.
+Until now!
+
+The setup was divided into two different categories: managed services wrappers and standard provisioning components.
+
+The first category is quite straight forward.
+For every managed service that is being used, a module is created.
+This module includes all the required Terraform resources for that service to operate.
+For example, the Key vault module contains the Azure Keyvault Terraform resource, but also the role assignments for the different Active Directory groups that require access to that Key vault.
```

Comment:
I'd write it like this to distinguish between the generic term and the Azure keyed term.
```suggestion
For example, the key vault module contains the Azure Key Vault Terraform resource, but also the role assignments for the different Active Directory groups that require access to that key vault.
```